service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementExtractor.java

```diff
@@ -0,0 +1,129 @@
+import uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil;
+import uk.nhs.adaptors.gp2gp.ehr.utils.TemplateUtils;
+
+public class MedicationStatementExtractor {
+    private static final Mustache IN_FULFILMENT_OF_TEMPLATE =
+        TemplateUtils.loadTemplate("in_fulfilment_of_template.mustache");
+
+    private static final String MEDICATION_QUANTITY_TEXT_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationQuantityText-1";
+    private static final String PRESCRIPTION_TYPE_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescriptionType-1";
+    private static final String REPEAT_INFORMATION_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationRepeatInformation-1";
+    private static final String NUM_OF_REPEAT_PRESCRIPTIONS_ALLOWED_URL = "numberOfRepeatPrescriptionsAllowed";
+    private static final String DEFAULT_REPEAT_VALUE = "1";
+    private static final String MEDICATION_STATUS_REASON_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationStatusReason-1";
+    private static final String STATUS_REASON_URL = "statusReason";
+    private static final String STATUS_CHANGE_URL = "statusChangeDate";
+    private static final String AVAILABILITY_TIME_VALUE_TEMPLATE = "<availabilityTime value=\"%s\"/>";
+    private static final String DEFAULT_QUANTITY_TEXT = "Unk UoM";
+
+    public static String extractDispenseRequestQuantityText(MedicationRequest medicationRequest) {
+        return medicationRequest.getDispenseRequest()
+            .getExtension()
+            .stream()
+            .filter(value -> MEDICATION_QUANTITY_TEXT_URL.equals(value.getUrl()))
+            .findFirst()
+            .map(value -> value.getValue().toString())
+            .orElse(DEFAULT_QUANTITY_TEXT);
+    }
+
+    public static String extractPrescriptionTypeCode(MedicationRequest medicationRequest) {
+        return filterExtensionByUrl(medicationRequest, PRESCRIPTION_TYPE_URL)
+            .map(Extension::getValue)
+            .map(CodeableConcept.class::cast)
+            .map(CodeableConcept::getCodingFirstRep)
+            .map(Coding::getCode)
+            .orElse(StringUtils.EMPTY);
+    }
+
+    public static String extractRepeatValue(MedicationRequest medicationRequest) {
+        var repeatInformation = filterExtensionByUrl(medicationRequest, REPEAT_INFORMATION_URL);
+
+        return repeatInformation.map(extension -> extension
+                .getExtension()
+                .stream()
+                .filter(value -> NUM_OF_REPEAT_PRESCRIPTIONS_ALLOWED_URL.equals(value.getUrl()))
+                .filter(Extension::hasValue)
+                .findFirst()
+                .map(BaseExtension::getValueAsPrimitive)
+                .map(IPrimitiveType::getValueAsString)
+                .orElse(DEFAULT_REPEAT_VALUE))
+            .orElse(DEFAULT_REPEAT_VALUE);
```

Comment:
Those duplicate lines at the end there indicate something fishy is going on. Rather than nesting the functional ops, we can probably get away with using a single chain
```suggestion
        return repeatInformation.map(Extension::getExtension).stream()
            .flatMap(List::stream)
            .filter(value -> NUM_OF_REPEAT_PRESCRIPTIONS_ALLOWED_URL.equals(value.getUrl()))
            .filter(Extension::hasValue)
            .findFirst()
            .map(BaseExtension::getValueAsPrimitive)
            .map(IPrimitiveType::getValueAsString)
            .orElse(DEFAULT_REPEAT_VALUE);
```